free_fleet/src/manager/Manager.cpp

```diff
@@ -0,0 +1,435 @@
+    return make_error_fn(
+      "Provided free fleet manager::CoordinateTransformer cannot be null.");
+  if (!time_now_fn)
+    return make_error_fn("Provided time function cannot be null.");
+
+  std::shared_ptr<Manager> manager_ptr(new Manager);
+  manager_ptr->_pimpl->fleet_name = fleet_name;
+  manager_ptr->_pimpl->graph = std::move(graph);
+  manager_ptr->_pimpl->middleware = std::move(middleware);
+  manager_ptr->_pimpl->to_robot_transform = std::move(to_robot_transform);
+  manager_ptr->_pimpl->time_now_fn = std::move(time_now_fn);
+  manager_ptr->_pimpl->robot_updated_callback_fn =
+    std::move(robot_updated_callback_fn);
+  return manager_ptr;
+}
+
+//==============================================================================
+Manager::Manager()
+: _pimpl(rmf_utils::make_impl<Implementation>(Implementation()))
+{
+}
+
+//==============================================================================
+void Manager::run_once()
+{
+  // Send out all unreceived tasks again
+  for (const auto t_it : _pimpl->unacknowledged_tasks)
+  {
+    t_it.second->send_request();
+  }
+}
+
+//==============================================================================
+auto Manager::robot_names() -> std::vector<std::string>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  std::vector<std::string> result;
+  result.reserve(_pimpl->robots.size());
+  for (const auto it : _pimpl->robots)
+  {
+    result.push_back(it.first);
+  }
+  return result;
+}
+
+//==============================================================================
+auto Manager::robot(const std::string& robot_name)
+-> std::shared_ptr<const manager::RobotInfo>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  const auto it = _pimpl->robots.find(robot_name);
+  if (it == _pimpl->robots.end())
+    return nullptr;
+  return it->second;
+}
+
+//==============================================================================
+auto Manager::all_robots()
+-> std::vector<std::shared_ptr<const manager::RobotInfo>>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  std::vector<std::shared_ptr<const manager::RobotInfo>> infos;
+  infos.reserve(_pimpl->robots.size());
+  for (const auto it : _pimpl->robots)
+  {
+    infos.push_back(it.second);
+  }
+  return infos;
+}
+
+//==============================================================================
+auto Manager::request_pause(const std::string& robot_name)
+-> std::optional<TaskId>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  if (_pimpl->robots.find(robot_name) == _pimpl->robots.end())
+    return std::nullopt;
+
+  const TaskId request_task_id = ++_pimpl->current_task_id;
+  messages::PauseRequest request(robot_name, request_task_id);
+
+  std::shared_ptr<manager::RequestInfo> request_info(
+    new manager::SimpleRequestInfo<messages::PauseRequest>(
+      request,
+      [this](const messages::PauseRequest& request_msg)
+      {
+        this->_pimpl->middleware->send_pause_request(request_msg);
```

Comment:
As mentioned in other places, it's risky to capture `this` in a lambda if you can't know when the lambda will be triggered.

Since the only thing that needs to be captured here is `middleware`, you could tweak the `middleware` field so that it's a `std::shared_ptr` instead of a `std::unique_ptr`, and then capture just `middleware` by copy here (or capture a `std::weak_ptr` to it; either should be fine).